### ISSUE TITLE
SSL context factory no longer reports errors at creation time

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3700]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3700|
|Type|release blocker: regression|
|Created|2009-03-23 04:15:47Z|
|Branch|https://github.com/twisted/twisted/tree/early-ssl-context-error-reporting-3700|

In Twisted 8.2, it was possible to detect bad arguments to `DefaultOpenSSLContextFactory` and `twisted.application.strports.parse` by catching `OpenSSL.SSL.Error`.

Now, due to the fact that `cacheContext()` is no longer called, these exceptions are no longer raised and you can't tell you've got a dud context until `getContext()` gets called.

You can see [SSL related failures on the Divmod Twisted-trunk buildbot caused by this](http://buildbot.divmod.org/builders/twisted-trunk/builds/133/steps/trial/logs/problems).

This regression was introduced in r25769, as a result of the changes for [#3330](https://github.com/twisted/twisted/issues/3330).

Even if we revert that revision, this behavior should be documented.

Here's a test program that demonstrates the problem:

```
#!py
from twisted.internet.ssl import DefaultOpenSSLContextFactory
DefaultOpenSSLContextFactory("no-private-key.pem", "no-cert.pem")
```

Under Twisted 8.2 this will raise an exception, under current trunk it will not.

<details><summary>Searchable metadata</summary>

```
trac-id__3700 3700
type__release_blocker__regression release blocker: regression
reporter__glyph glyph
priority__highest highest
milestone__ 
branch__branches_early_ssl_context_error_reporting_3700 branches/early-ssl-context-error-reporting-3700
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1237781747000000 1237781747000000
changetime__1237920658000000 1237920658000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [26521]) Branching to 'early-ssl-context-error-reporting-3700'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Previous behavior restored, with unit tests.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @exarkun<br>@glyph set status to `new`|
|-|-|

Nice clean fix.  Please add documentation for the `_contextFactory` attribute/parameter to `DefaultOpenSSLContextFactory` and land.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [26524]) Give DefaultOpenSSLContextFactory a class docstring and document _contextFactory in it

refs [#3700](https://github.com/twisted/twisted/issues/3700)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [26525]) Merge early-ssl-context-error-reporting-3700 

Author: exarkun
Reviewer: glyph
Fixes: [#3700](https://github.com/twisted/twisted/issues/3700)

Reinstate the early failure mode for invalid parameters passed to
`DefaultOpenSSLContextFactory` and add tests for this behavior so
it is preserved in the future.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
